catalog/harvest/catalog-harvest-file/src/main/java/org/codice/ddf/catalog/harvest/file/DirectoryHarvester.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.commons.io.monitor.FileAlterationObserver;
+import org.codice.ddf.catalog.harvest.Listener;
+import org.codice.ddf.catalog.harvest.common.FileSystemPersistenceProvider;
+import org.codice.ddf.catalog.harvest.common.PollingHarvester;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DirectoryHarvester extends PollingHarvester {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DirectoryHarvester.class);
+
+  private static final long DEFAULT_POLL_INTERVAL = 5;
+
+  private final Set<Listener> listeners = new HashSet<>();
+
+  private final File harvestFile;
+
+  private final FileSystemPersistenceProvider fileSystemPersistenceProvider;
+
+  private final FileAlterationObserver fileAlterationObserver;
+
+  private final FileAlterationListenerAdaptor fileListener;
+
+  private final String persistenceKey;
+
+  /**
+   * Creates and starts a {@link org.codice.ddf.catalog.harvest.Harvester} that harvests a
+   * directory.
+   *
+   * @param dir an absolute path to a directory
+   */
+  public DirectoryHarvester(String dir) {
+    this(dir, Collections.emptySet());
+  }
+
+  /**
+   * Creates and starts a {@link org.codice.ddf.catalog.harvest.Harvester} that harvests a
+   * directory.
+   *
+   * @param dir an absolute path to a directory
+   * @param listeners initial set of {@link Listener}s
+   */
+  public DirectoryHarvester(String dir, Set<Listener> listeners) {
+    this(dir, DEFAULT_POLL_INTERVAL, listeners);
+  }
+
+  public DirectoryHarvester(String dir, long pollInterval, Set<Listener> inititialListeners) {
+    super(pollInterval);
+    harvestFile = new File(dir);
+    validateDirectory(dir);
+    persistenceKey = DigestUtils.sha1Hex(dir);
+
+    inititialListeners.forEach(this::registerListener);
+    fileSystemPersistenceProvider = new FileSystemPersistenceProvider("harvest/directory");
+    fileAlterationObserver = getCachedObserverOrCreate(persistenceKey, dir);
+    fileListener = new DirectoryHarvesterListenerAdaptor(listeners);
+
+    super.init();
+  }
+
+  private FileAlterationObserver getCachedObserverOrCreate(String key, String dir) {
+    if (fileSystemPersistenceProvider.loadAllKeys().contains(key)) {
+      LOGGER.trace("existing file observer for persistence key [{}] found, loading observer", key);
+      return (FileAlterationObserver) fileSystemPersistenceProvider.loadFromPersistence(key);
+    }
+
+    LOGGER.trace("no existing file observer for persistence key [{}], creating new observer", key);
+    return new FileAlterationObserver(dir);
+  }
+
+  @Override
+  public void poll() {
+    fileAlterationObserver.addListener(fileListener);
+    fileAlterationObserver.checkAndNotify();
+    fileAlterationObserver.removeListener(fileListener);
+    fileSystemPersistenceProvider.store(persistenceKey, fileAlterationObserver);
+  }
+
+  @Override
+  public void registerListener(Listener listener) {
+    if (!listeners.isEmpty()) {
```

Comment:
❓If we can use the Apache Commons' `EventListenerSupport` class, should we just ignore this check and allow multiple listeners?

Reply:
❓Instead of adding and removing listeners here and doing the notifications in the `DirectoryHarvesterListenerAdaptor` class, would it make more sense to move the `registerListener` and `unregisterListener` methods to the `DirectoryHarvesterListenerAdaptor` class and simply delegate down to that class here? That way all the listener management code would be in one class and we wouldn't need to pass the listeners' list around.